fw/http_frame.c

```diff
@@ -991,31 +996,62 @@ tfw_h2_headers_process(TfwH2Ctx *ctx)
 	return tfw_h2_stream_state_process(ctx);
 }
 
+static int
+tfw_h2_increment_wnd_sz(long int *window, unsigned int wnd_incr)
+{
+	long int new_window = *window + wnd_incr;
+	/*
+	 * According to RFC 9113 6.9.1
+	 * A sender MUST NOT allow a flow-control window to exceed 2^31-1 octets.
+	 * If a sender receives a WINDOW_UPDATE that causes a flow-control
+	 * window to exceed this maximum, it MUST terminate either the stream
+	 * or the connection, as appropriate. For streams, the sender sends a
+	 * RST_STREAM with an error code of FLOW_CONTROL_ERROR; for the
+	 * connection, a GOAWAY frame with an error code of FLOW_CONTROL_ERROR
+	 * is sent.
+	 */
+	if (new_window > MAX_WND_SIZE)
+		return T_DROP;
+	*window = new_window;
+	return 0;
+}
+
 static int
 tfw_h2_wnd_update_process(TfwH2Ctx *ctx)
 {
 	unsigned int wnd_incr;
 	TfwFrameHdr *hdr = &ctx->hdr;
+	TfwH2Err err_code = HTTP2_ECODE_PROTO;
 
 	wnd_incr = ntohl(*(unsigned int *)ctx->rbuf) & ((1U << 31) - 1);
 	if (wnd_incr) {
-		/*
-		 * TODO #498: apply new window size for entire connection or
-		 * particular stream.
-		 */
+		TfwH2Conn *conn = container_of(ctx, TfwH2Conn, h2);
+		long int *window = ctx->cur_stream ?
+			&ctx->cur_stream->rem_wnd : &ctx->rem_wnd;
+		int size, mss;
+
+		if (tfw_h2_increment_wnd_sz(window, wnd_incr)) {
+			err_code = HTTP2_ECODE_FLOW;
+			goto fail;
+		}
+
+		mss = tcp_send_mss(((TfwConn *)conn)->sk, &size, MSG_DONTWAIT);
+		tcp_push(((TfwConn *)conn)->sk, MSG_DONTWAIT, mss,
```

Comment:
Is there a sense to call `tcp_push` in cases when `window` is too small, for instance `window < FRAME_HEADER_SIZE` or even `window` is negative?

Reply:
You right we should not call tcp_push for negative window, but for window < FRAME_HEADER_SIZE it is ok, because frame header doesn't occur in flow control.